### PR TITLE
Grocery list

### PR DESCRIPTION
**Pull request by SeanT18**
I implement the ability to add groceries and display them on the grocery page. In my last commit There is a lot of files because I  pull from master during that time and forgot to commit that merge separately instead. The files and code that relate to tasks are already in master so don't pay attention to those.  

I tested this by entering the grocery page and adding values to the list and have them be shown

### CODE REVIEW

**Comment by SeanT18** [comment]
I thought it would show all of the files that I merged from master but it didn't so that's good. Also there's probably a way to do the grocery and tasks in the same method in the js file. I'm choosing to worry about that in the mvp instead.

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        
+        //loads entities into arraylist to be printed
+        for (Entity entity : results.asIterable()) {
+            String groceryItem = (String) entity.getProperty("grocery");
+            groceryList.add(groceryItem);
+        }
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(groceryList));
+        }
```
Comment (sidharthp) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Fix the indentation of this.

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/** Servlet responsible for creating new tasks. */
```
Comment (sidharthp) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Update the comment.

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        // creates arraylist and starts query
+        ArrayList<String> groceryList = new ArrayList<String>();
+        Query query = new Query("grocery");
+        PreparedQuery results = datastore.prepare(query);
+        
+        //loads entities into arraylist to be printed
+        for (Entity entity : results.asIterable()) {
+            String groceryItem = (String) entity.getProperty("grocery");
+            groceryList.add(groceryItem);
+        }
```
Comment (sidharthp) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Maybe I'm understanding this incorrectly, but this entity has no information about what family it belongs to. Is it shared across all families?

File: src/main/webapp/script.js [hunk 4]
```diff
@@ -33,9 +33,39 @@ function userLogin() {
...
     document.getElementById('login-container').innerHTML = message;
   });
-  
+}
+/*function loadGrocery() {
+  fetch('/grocery-list').then(response => response.text())
+  .then((message) => {
+    document.getElementById('grocery-list-container');
+  });
+}*/
```
Comment (sidharthp) [inline src/main/webapp/script.js]: Delete commented code.

File: src/main/webapp/script.js [hunk 5]
```diff
@@ -33,9 +33,39 @@ function userLogin() {
...
+
+    const titleElement = document.createElement('span');
+    titleElement.innerText = grocery;
+
+    groceryElement.appendChild(titleElement);
+    return groceryElement;
+}
+
+
+
```
Comment (sidharthp) [inline src/main/webapp/script.js]: Only keep 1 empty line between functions.

File: src/main/webapp/script.js [hunk 6]
```diff
@@ -33,9 +33,39 @@ function userLogin() {
...
+}*/
+
+
+function loadGrocery() {
+    // fetches json list of groceries
+    fetch('/grocery-list').then(response => response.json()).then((groceries) => {
+        const groceryListElement = document.getElementById('grocery-list-container');
+      
+    groceries.forEach((grocery) => {
+        groceryListElement.appendChild(createGroceryElement(grocery));
```
Comment (sidharthp) [inline src/main/webapp/script.js:44]: Fix indentation.

**Review by sidharthp — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        doGet(request,response);  
+        response.sendRedirect("/grocery.html");
+    }
+        public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        // adds item to datastore
+        String grocery = request.getParameter("groceryItem");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity groceryEntity = new Entity("grocery");
+        if(grocery != null && !grocery.equals("")) {
```
Comment (yjennywang) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: if (

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+/** Servlet responsible for creating new tasks. */
+@WebServlet("/grocery-list")
+public class GroceryServlet extends HttpServlet {
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        doGet(request,response);  
+        response.sendRedirect("/grocery.html");
+    }
+        public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (yjennywang) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: don't do datastore puts in a doGet

In general:

- GET requests are for fetching data to return to the page (with no side effects)
- POST requests are for side effects (e.g. submitting a form, creating/updating/deleting elements)

Would it make more sense to separate the "add a grocery item" and "list all groceries" http methods? Can you also sync up with Apoorva (who added thet asks) to make sure things are consistent?

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        // adds item to datastore
+        String grocery = request.getParameter("groceryItem");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity groceryEntity = new Entity("grocery");
```
Comment (yjennywang) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Can you ensure that the capitalization and parameter name styles here are consistent with the capitalization and parameter names that Apoorva and Anthony have been writing?

Bonus points for creating a style guide doc for you all to follow. :)

Comment (yjennywang) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Not a blocker right now, but https://github.com/googleinterns/step122-2020/issues/9 applies here too.

**Comment by yjennywang** [comment]
You may want to also consider pulling tasks/groceries into a separate javascript file (tasks.js) which you can then include.

This means you  would only need to include it in the HTML files where it is used. And script.js will grow very quickly, and it will be harder to separate later.

**Comment by yjennywang** [comment]
Can you add how you have tested the code in the PR description?

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 1 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: solved

File: src/main/webapp/script.js (hunk 4 above)
Comment (SeanT18) [reply src/main/webapp/script.js]: solved

File: src/main/webapp/script.js (hunk 6 above)
Comment (SeanT18) [reply src/main/webapp/script.js:44]: solved

File: src/main/webapp/script.js (hunk 5 above)
Comment (SeanT18) [reply src/main/webapp/script.js]: solved

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 3 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: currently it is not. I thought that function would be something we'd work on in the alpha

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 7 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: solved

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 8 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: so by making things consistent do you mean making three separate servlets like she did? or something else

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 3 above)
Comment (sidharthp) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: Add a TODO then in the code to make sure that you make that change.

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 2 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: solved

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 3 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: will do

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 8 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: solved

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 9 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: solved I think

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+
+/** Servlet responsible for creating new tasks. */
+@WebServlet("/grocery-list")
+public class GroceryServlet extends HttpServlet {
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      doGet(request,response);  
+      response.sendRedirect("/grocery.html");
+    }
```
Comment (apoorvadhanala) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: add a line between the methods

File: src/main/webapp/script.js [hunk 11]
```diff
@@ -33,7 +33,28 @@ function userLogin() {
...
+      
+    groceries.forEach((grocery) => {
+      groceryListElement.appendChild(createGroceryElement(grocery));
+    })
+  });
+}
+
+function createGroceryElement(grocery){
+    const groceryElement = document.createElement('li');
+      groceryElement.className = 'task';
```
Comment (apoorvadhanala) [inline src/main/webapp/script.js]: fix indent

**Review by apoorvadhanala — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/GroceryServlet.java (hunk 10 above)
Comment (SeanT18) [reply src/main/java/com/google/sps/servlets/GroceryServlet.java]: solved

File: src/main/webapp/script.js (hunk 11 above)
Comment (SeanT18) [reply src/main/webapp/script.js]: solved

**Review by SeanT18 — COMMENTED** [review]
For this push I actually just fixed styling errors. I thought I didn't push the function that I titled it but I did.

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet takes grocery list and displays it to the screen based on the users family*/
```
Comment (sidharthp) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: nit: Add space between `family` and `*/`

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity groceryEntity = new Entity("Grocery");
+    if( grocery != null && !grocery.equals("")) {
+        groceryEntity.setProperty("grocery", grocery);
+        datastore.put(groceryEntity);
+    }
+               
+    doGet(request,response);  
+    response.sendRedirect("/grocery.html");
+    }
```
Comment (sidharthp) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Fix indentation

File: src/main/java/com/google/sps/servlets/GroceryServlet.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    PreparedQuery results = datastore.prepare(query);
+        
+    //loads entities into arraylist to be printed
+    for (Entity entity : results.asIterable()) {
+        String groceryItem = (String) entity.getProperty("grocery");
+        groceryList.add(groceryItem);
+    }
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(groceryList));
```
Comment (sidharthp) [inline src/main/java/com/google/sps/servlets/GroceryServlet.java]: Indent +2

**Review by anthonyqzhu — APPROVED** [review]

**Review by apoorvadhanala — APPROVED** [review]

**Pull request merged**